Add docs to a separate PR for review (#5153)
ROCm/rocm-libraries: projects/hipdnn/docs/conceptual/architecture.rst

.. meta::
  :description: hipDNN has a plugin-based architecture in order to allow contributors and users to extend hipDNN without modifying the core library. 
  :keywords: hipDNN, ROCm, API, 

.. _architecture:

******************************
hipDNN high-level architecture
******************************

hipDNN has a plugin-based architecture in order to allow contributors and users to extend hipDNN without modifying the core library. 

> hipDNN has support for engine plugins, which provide the kernels to solve graphs. 

Reviewer: The graphic below may be easier to digest by readers if the description of each component was provided before. The following is largely copied-and-pasted from the "Components" section below, but with some additional explanation.

The Hipdnn library is structured as three primary components:
- **Frontend**: The hipDNN Frontend is a header-only C++ library that provides an industry standard API for interacting with hipDNN. The frontend wraps the backend C API to provide a more user-friendly C++ interface, encapsulating much of the detail required when creating graphs using the backend API.
- **Backend**: The hipDNN Backend is a shared library which provides a C API for hipDNN. The backend is the core component of hipDNN which acts as a plugin loader and manager, connecting problems to engines optimized to solve them.
- **Engine Plugins** The hipDNN Engine Provider Plugins are shared libraries that are responsible for matching the operations they provide to graphs and executing the operations on the supported hardware. Plugins will continue to be added over time to provide additional operational support or performance improvements. See :ref:`plugin-support` for more information.
- **Heuristic & Selection Plugins (Future)** The Engine heuristic and selection plugins will be used to add increased flexibility in the matching graphs with the most suitable engine.

The Frontend API is used to define tensors and attach them to operational nodes on a graph. The graph is then lowered through the backend APIs where it is examined by each plugin's engine to determine a match., at which point n execution plan is established with the preferred matching engine. The frontend then allocates memory and populates the data used by the graph, which is then dispatched to the selected plugin's engine to run.

Header-only SDKs provide shared utilities and interfaces. hipDNN provides three SDKs: Data SDK (graph schemas and data structures), Plugin SDK (plugin API and utilities), and Test SDK (testing utilities and CPU reference implementations).

Author: Added. However, if the Heuristic plugins aren't available yet, we shouldn't mention it yet.

.. note::

  See :ref:`backend-architecture` for a more granular breakdown of the system architecture and the backend API.

Architecture overview
=====================

.. image:: ../images/hipDNN_Architecture.png

> Components

Reviewer: Moved to the top of this page.

==========

- **Frontend**: A header-only C++ library that provides the industry standard API for interacting with hipDNN. The frontend wraps the backend C API to provide a more user-friendly C++ interface.
- **Backend**: A shared library which provides a C API for hipDNN. The backend is the core component of hipDNN which acts as a plugin loader and manager, connecting problems to plugins that can solve them.
- **SDKs**: Header-only libraries that provide shared utilities and interfaces. hipDNN provides three SDKs: Data SDK (graph schemas and data structures), Plugin SDK (plugin API and utilities), and Test SDK (testing utilities and CPU reference implementations).
- **MIOpen Provider Plugin**: A plugin that wraps MIOpen and provides access to the existing API through hipDNN. The :ref:`miopen` is its own separate project from hipDNN.
- **Other Plugins**: Plugins will be added over time to provide additional operational support, or performance improvements. Plugins should be external projects to hipDNN. See :ref:`plugin-support` for more information.

> Frontend

Reviewer: This and what follows for the Frontend and Backend components on this page LGTM to remain here.

--------

The frontend provides a user-friendly C++ interface to hipDNN, wrapping the lower-level C API provided by the backend.

Key characteristics:

- **Header-only C++ library**: There are no compiled libraries, which simplifies integration.
- **Dependencies**: The frontend is dependent on the :ref:`backend` and :ref:`data`.
- **Purpose**: It provides the API for users consuming hipDNN.
- **Expected usage**: The frontend should be consumed as a header-only dependency in user projects.

Frontend architecture
~~~~~~~~~~~~~~~~~~~~~

Graph class
^^^^^^^^^^^

The central abstraction in the frontend is the ``Graph`` class, which:

- Manages the construction of operation graphs.
- Handles the creation and configuration of nodes.
- Orchestrates the execution workflow.

Nodes
^^^^^

Nodes represent individual operations within a graph:

- Each node type (for example, ``BatchnormNode``, ``PointwiseNode``) inherits from ``INode``.
- Nodes encapsulate their specific attributes and tensor connections.

> - Support serialization to Flatbuffer format for backend consumption.

Reviewer: This line is no longer correct (the code is in the process of being changed).

```suggestion
- Uses the Backend API to convert the graph to backend descriptors for backend consumption.
```

Attributes
^^^^^^^^^^

Attributes configure the behavior of nodes:

- Each node type has corresponding attribute classes (for example, ``Batchnorm_attributes``).
- Attributes include operation-specific parameters like epsilon, momentum, etc.

> - Support builder pattern for easy configuration.

Reviewer: This line isn't helpful for a User. It can be removed.

> 

Reviewer: TensorAttributes
^^^^^^^^^^^^^^^^

Tensors attributes define the shape of the data processed by operations in the graph. 
- Dimensions of the data.
- Memory layout of the data (e.g., NCHW, NHWC), controlled by strides.
- The data's type
- Attached to the graph operation nodes.

Author: Added

Reviewer: Hi @mattwill-amd. Reviewing this with the team, This should be "TensorAttributes" instead of "Tensor". I've updated the heading and first sentence above accordingly. Sorry for the rework on this one.

Reviewer: UPDATE: also changed "How the data is packed in memory." to "Memory layout of the data (e.g., NCHW, NHWC), controlled by strides."

Simplified workflow example
^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code:: cpp

  // Create a graph
  Graph graph;
  graph.set_compute_data_type(DataType_t::FLOAT);

  // Create tensors
  auto x = Graph::tensor(/* tensor attributes */);
  auto scale = Graph::tensor(/* tensor attributes */);
  auto bias = Graph::tensor(/* tensor attributes */);

  // Add operations
  auto [y, mean, inv_var, _, _] = graph.batchnorm(x, scale, bias, bn_attributes);

  // Build and execute
  graph.build_operation_graph(handle);
  graph.create_execution_plans();
  graph.build_plans();
  graph.execute(handle, variant_pack, workspace);

For complete working examples, see the official `samples on GitHub <https://github.com/ROCm/rocm-libraries/tree/develop/projects/hipdnn/samples>`_.

SDKs

> ----

Reviewer: The SDK and Plugin sections don't describe architecture that is relevant to Users; it's relevant to Plugin Developers.

It could be moved to an "Architecture for Plugin Developers" page (following the Install for Plugin Developers" pattern).

Author: Separated the sections between users and plugin developers for now.

hipDNN provides three header-only SDK libraries that serve as the foundation for communication between different components.

.. _data:

> Data SDK (``data_sdk``)

Reviewer: The remaining content in this document doesn't describe architecture that is relevant to Users; it's relevant to Plugin Developers.

It could be moved to an "Architecture for Plugin Developers" page (following the Install for Plugin Developers" pattern).

~~~~~~~~~~~~~~~~~~~~~~~

The Data SDK contains ``FlatBuffers`` schemas and data structures for graph representation. 

> The serialized structures allow data to be marshalled and passed between the frontend, backend, and plugins in a type-safe and highly version-compatible fashion.

Reviewer: ```suggestion
The serialized structures allow data to be marshalled and passed between the backend and plugins in a type-safe and highly version-compatible fashion.
```

- **Dependencies**: ``FlatBuffers`` and ``spdlog``.
- **Purpose**: Provides data structures and serialization for graphs, tensors, and configurations.

> - **Expected usage**: Consumed by the frontend, backend, and plugins for graph data handling.

Reviewer: ```suggestion
- **Expected usage**: Consumed by the backend and plugins for graph data handling.
```

  - Test utilities (tolerances, seeds, logging).
  - Mock objects for unit testing.

Plugin architecture
-------------------

Plugin loading
~~~~~~~~~~~~~~~

- The backend discovers plugins at runtime via the default plugin path, or by using ``hipdnnSetEnginePluginPaths_ext`` to provide additional paths from which to load the plugins.
- Each plugin exports standard entry points defined in the Plugin SDK.

Engine management
~~~~~~~~~~~~~~~~~

- Each plugin can provide multiple engines.
- Engines must have globally unique IDs that remain constant for each run.
- Plugins determine which engines are applicable for a given graph.

Key plugin functions
~~~~~~~~~~~~~~~~~~~~

.. code:: c

  // Get all available engine IDs
  hipdnnEnginePluginGetAllEngineIds(engine_ids, max_engines, num_engines);

  // Check which engines can solve a graph
  hipdnnEnginePluginGetApplicableEngineIds(handle, graph, engine_ids, max, num);

  // Create execution context for a specific engine
  hipdnnEnginePluginCreateExecutionContext(handle, config, graph, context);

  // Execute the graph
  hipdnnEnginePluginExecuteOpGraph(handle, context, workspace, buffers, num_buffers);

> Engine plugins

Reviewer: ```suggestion
Plugin engines
```
This section could be moved above Engine management so that readers learn what an engine is before they learn about how they are managed.

--------------

> Engine plugins provide the actual computational implementations for hipDNN graphs.

Reviewer: ```suggestion
Engines, provided by each plugin, provide the actual computational implementations for hipDNN graphs.
```

Key characteristics:

- **Separate installable projects**: Independent development and deployment.
- **Dependencies**: :ref:`data` and :ref:`plugin-sdk` (and plugin-specific dependencies as needed).
- **Purpose**: Provides engines which are capable of solving graphs.
- **Expected usage**: Loaded at runtime by the backend.

> Engine plugin types

Reviewer: ```suggestion
Engine types
```

5. **Run execution plan**: Execute with variant pack containing tensor data.

.. code:: c

  // Simplified Backend workflow
  hipdnnBackendDescriptor_t graph_desc, heuristic_desc, config_desc, plan_desc, variant_desc;

  // 1. Create graph (from serialized data)
  hipdnnBackendCreateAndDeserializeGraph_ext(&graph_desc, serialized_graph, size);

  // 2. Create and configure heuristic
  hipdnnBackendCreateDescriptor(HIPDNN_BACKEND_ENGINEHEUR_DESCRIPTOR, &heuristic_desc);
  hipdnnBackendSetAttribute(heuristic_desc, HIPDNN_ATTR_ENGINEHEUR_OPERATION_GRAPH, ...);
  hipdnnBackendFinalize(heuristic_desc);

  // 3. Get engine configurations
  hipdnnBackendGetAttribute(heuristic_desc, HIPDNN_ATTR_ENGINEHEUR_RESULTS, ...);

  // 4. Create execution plan
  hipdnnBackendCreateDescriptor(HIPDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR, &plan_desc);
  hipdnnBackendSetAttribute(plan_desc, HIPDNN_ATTR_EXECUTION_PLAN_ENGINE_CONFIG, ...);
  hipdnnBackendFinalize(plan_desc);

  // 5. Execute
  hipdnnBackendExecute(handle, plan_desc, variant_desc);

Error handling strategy
-----------------------

hipDNN uses a layered error handling approach designed to be robust across C/C++ boundaries:

> -  **Plugins**: Plugin entry points return ``hipdnnPluginStatus_t`` codes. Internal exceptions are caught at the plugin boundary and converted to status codes. Error strings are stored in thread-local storage via ``PluginLastErrorManager``.

Reviewer: Move this point to the Plugin Developer architecture page.

Author: Done

-  **Backend (C API)**: All public API functions return ``hipdnnStatus_t`` codes. The backend catches any internal C++ exceptions, converts them to the appropriate status code, and stores the exception message. Users can retrieve descriptive error messages using ``hipdnnGetLastErrorString``.

> -  **Frontend (C++ API)**: The C++ frontend checks ``hipdnnStatus_t`` codes from the backend. On failure, it retrieves the detailed error message via ``hipdnnGetLastErrorString`` and returns an ``Error`` object containing the error code and description. The frontend utilizes *value-based error handling* rather than throwing exceptions.

Reviewer: Would flow more naturally with the layout of this page if Frontend was listed first, before backend.
